App force quits at startup

This appears to be due to a feature which an unrecognized layerId in the local database.

Since Firestore is a schemaless db, the app needs to be tolerant to invalid data in the remote db. Invalid data likely ended up getting synced to the local db, causing this exception.

Stack trace:

```
2020-11-25 11:09:34.207 24805-24941/com.google.android.gnd E/eglCodecCommon: GoldfishAddressSpaceHostMemoryAllocator: ioctl_ping failed for device_type=5, ret=-1
2020-11-25 11:09:35.018 24805-24805/com.google.android.gnd E/SchedPolicy: set_timerslack_ns write failed: Operation not permitted
2020-11-25 11:09:37.346 24805-24805/com.google.android.gnd E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.google.android.gnd, PID: 24805
    java.lang.RuntimeException: LiveData does not handle errors. Errors from publishers should be handled upstream and propagated as state
        at androidx.lifecycle.LiveDataReactiveStreams$PublisherLiveData$LiveDataSubscriber$1.run(LiveDataReactiveStreams.java:264)
        at android.os.Handler.handleCallback(Handler.java:873)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:193)
        at android.app.ActivityThread.main(ActivityThread.java:6669)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
     Caused by: java.util.NoSuchElementException: No value present
        at java8.util.Optional.orElseThrow(Optional.java:376)
        at java8.util.Optional.get(Optional.java:143)
        at com.google.android.gnd.persistence.local.room.entity.FeatureEntity.toFeature(FeatureEntity.java:120)
        at com.google.android.gnd.persistence.local.room.RoomLocalDataStore.lambda$getFeaturesOnceAndStream$10(RoomLocalDataStore.java:256)
        at com.google.android.gnd.persistence.local.room.-$$Lambda$RoomLocalDataStore$3URsonvOsxXBcz_pc3-Ofs0U3e8.apply(Unknown Source:4)
        at java8.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:196)
        at java8.util.DelegatingSpliterator$ConsumerDelegate.accept(DelegatingSpliterator.java:93)
        at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1390)
        at java8.util.DelegatingSpliterator.forEachRemaining(DelegatingSpliterator.java:57)
        at java8.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:488)
        at java8.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:472)
        at java8.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:1269)
        at java8.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java8.util.stream.ReferencePipeline.collect(ReferencePipeline.java:602)
        at com.google.android.gnd.persistence.local.room.RoomLocalDataStore.lambda$getFeaturesOnceAndStream$11(RoomLocalDataStore.java:257)
        at com.google.android.gnd.persistence.local.room.-$$Lambda$RoomLocalDataStore$-x2oAeF-36xO3LZNOf9Yvvud7hU.apply(Unknown Source:4)
        at io.reactivex.internal.operators.flowable.FlowableMap$MapSubscriber.onNext(FlowableMap.java:63)
        at com.akaita.java.rxjava2debug.extensions.FlowableOnAssembly$OnAssemblySubscriber.onNext(FlowableOnAssembly.java:61)
        at io.reactivex.internal.operators.flowable.FlowableFlatMapMaybe$FlatMapMaybeSubscriber.innerSuccess(FlowableFlatMapMaybe.java:175)
        at io.reactivex.internal.operators.flowable.FlowableFlatMapMaybe$FlatMapMaybeSubscriber$InnerObserver.onSuccess(FlowableFlatMapMaybe.java:397)
        at com.akaita.java.rxjava2debug.extensions.MaybeOnAssembly$OnAssemblyMaybeObserver.onSuccess(MaybeOnAssembly.java:68)
        at io.reactivex.internal.operators.maybe.MaybeFromCallable.subscribeActual(MaybeFromCallable.java:61)
        at io.reactivex.Maybe.subscribe(Maybe.java:4290)
        at com.akaita.java.rxjava2debug.extensions.MaybeOnAssemblyCallable.subscribeActual(MaybeOnAssemblyCallable.java:43)
        at io.reactivex.Maybe.subscribe(Maybe.java:4290)
        at io.reactivex.internal.operators.flowable.FlowableFlatMapMaybe$FlatMapMaybeSubscriber.onNext(FlowableFlatMapMaybe.java:132)
        at com.akaita.java.rxjava2debug.extensions.FlowableOnAssembly$OnAssemblySubscriber.onNext(FlowableOnAssembly.java:61)
        at io.reactivex.internal.operators.flowable.FlowableObserveOn$ObserveOnSubscriber.runAsync(FlowableObserveOn.java:407)
        at io.reactivex.internal.operators.flowable.FlowableObserveOn$BaseObserveOnSubscriber.run(FlowableObserveOn.java:176)
        at io.reactivex.internal.schedulers.ExecutorScheduler$ExecutorWorker$BooleanRunnable.run(ExecutorScheduler.java:288)
        at io.reactivex.internal.schedulers.ExecutorScheduler$ExecutorWorker.run(ExecutorScheduler.java:253)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
2020-11-25 11:09:37.347 24805-24805/com.google.android.gnd E/AndroidRuntime:     at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:764)
     Caused by: com.akaita.java.rxjava2debug.extensions.RxJavaAssemblyException
     Caused by: com.akaita.java.rxjava2debug.extensions.RxJavaAssemblyException
     Caused by: com.akaita.java.rxjava2debug.extensions.RxJavaAssemblyException
     Caused by: com.akaita.java.rxjava2debug.extensions.RxJavaAssemblyException
2020-11-25 11:09:37.181 24805-24805/com.google.android.gnd E/SchedPolicy: set_timerslack_ns write failed: Operation not permitted
```

--- a/gnd/src/main/java/com/google/android/gnd/persistence/local/room/RoomLocalDataStore.java
+++ b/gnd/src/main/java/com/google/android/gnd/persistence/local/room/RoomLocalDataStore.java
@@ -85,6 +85,7 @@ import io.reactivex.Single;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.List;
+import java8.util.Optional;
 import javax.inject.Inject;
 import javax.inject.Singleton;
 import timber.log.Timber;
@@ -144,8 +145,9 @@ public class RoomLocalDataStore implements LocalDataStore {
             Observable.just(field)
                 .filter(__ -> field.getMultipleChoice() != null)
                 .flatMapCompletable(
-                    __ -> insertOrUpdateMultipleChoice(field.getId(),
-                        checkNotNull(field.getMultipleChoice()))))
+                    __ ->
+                        insertOrUpdateMultipleChoice(
+                            field.getId(), checkNotNull(field.getMultipleChoice()))))
         .subscribeOn(schedulers.io());
   }
 
@@ -250,12 +252,16 @@ public class RoomLocalDataStore implements LocalDataStore {
   public Flowable<ImmutableSet<Feature>> getFeaturesOnceAndStream(Project project) {
     return featureDao
         .findOnceAndStream(project.getId(), EntityState.DEFAULT)
-        .map(
-            list ->
-                stream(list)
-                    .map(f -> FeatureEntity.toFeature(f, project))
-                    .collect(toImmutableSet()))
+        .map(featureEntities -> toFeatures(project, featureEntities))
         .subscribeOn(schedulers.io());
+  }
+
+  private ImmutableSet<Feature> toFeatures(Project project, List<FeatureEntity> featureEntities) {
+    return stream(featureEntities)
+        .map(f -> FeatureEntity.toFeature(f, project))
+        .filter(f -> f.isPresent())
+        .map(Optional::get)
+        .collect(toImmutableSet());
   }
 
   // TODO(#127): Decouple from Project and remove project from args.
@@ -264,6 +270,8 @@ public class RoomLocalDataStore implements LocalDataStore {
     return featureDao
         .findById(featureId)
         .map(f -> FeatureEntity.toFeature(f, project))
+        .filter(Optional::isPresent)
+        .map(Optional::get)
         .subscribeOn(schedulers.io());
   }
 

--- a/gnd/src/main/java/com/google/android/gnd/persistence/local/room/entity/FeatureEntity.java
+++ b/gnd/src/main/java/com/google/android/gnd/persistence/local/room/entity/FeatureEntity.java
@@ -27,10 +27,13 @@ import com.google.android.gnd.model.AuditInfo;
 import com.google.android.gnd.model.Project;
 import com.google.android.gnd.model.feature.Feature;
 import com.google.android.gnd.model.feature.FeatureMutation;
+import com.google.android.gnd.model.layer.Layer;
 import com.google.android.gnd.persistence.local.room.models.Coordinates;
 import com.google.android.gnd.persistence.local.room.models.EntityState;
 import com.google.auto.value.AutoValue;
 import com.google.auto.value.AutoValue.CopyAnnotations;
+import java8.util.Optional;
+import timber.log.Timber;
 
 /**
  * Defines how Room persists features in the local db. By default, Room uses the name of object
@@ -113,16 +116,22 @@ public abstract class FeatureEntity {
   }
 
   // TODO(#127): Decouple from Project and remove 2nd argument.
-  public static Feature toFeature(FeatureEntity featureEntity, Project project) {
-    return Feature.newBuilder()
-        .setId(featureEntity.getId())
-        .setProject(project)
-        .setLayer(project.getLayer(featureEntity.getLayerId()).get())
-        .setPoint(featureEntity.getLocation().toPoint())
-        .setGeoJsonString(featureEntity.getGeoJson())
-        .setCreated(AuditInfoEntity.toObject(featureEntity.getCreated()))
-        .setLastModified(AuditInfoEntity.toObject(featureEntity.getLastModified()))
-        .build();
+  public static Optional<Feature> toFeature(FeatureEntity featureEntity, Project project) {
+    Optional<Layer> layer = project.getLayer(featureEntity.getLayerId());
+    if (layer.isEmpty()) {
+      Timber.d("Skipping feature " + featureEntity.getId() + " with unknown layerId");
+      return Optional.empty();
+    }
+    return Optional.of(
+        Feature.newBuilder()
+            .setId(featureEntity.getId())
+            .setProject(project)
+            .setLayer(layer.get())
+            .setPoint(featureEntity.getLocation().toPoint())
+            .setGeoJsonString(featureEntity.getGeoJson())
+            .setCreated(AuditInfoEntity.toObject(featureEntity.getCreated()))
+            .setLastModified(AuditInfoEntity.toObject(featureEntity.getLastModified()))
+            .build());
   }
 
   public abstract FeatureEntity.Builder toBuilder();

